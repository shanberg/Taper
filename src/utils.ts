export function formatDate(date, style = 'conciseUS') {
let options;

  switch (style) {
    case 'conciseUS':
      options = { month: 'numeric', day: 'numeric' };
      break;
    case 'conciseEU':
      options = { day: 'numeric', month: 'numeric', year: '2-digit' };
      break;
    case 'longform':
      options = { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' };
      break;
    // Add more styles as needed
    default:
      options = {};
  }

  return new Intl.DateTimeFormat('default', options).format(new Date(date));
}

export const isRowValid = (row): boolean => {
  return row.dose <= 0 || row.daysForDose <= 0
}

export const isRowPlaceholder = (row): boolean => {
  return row.dose === 0 && row.daysForDose === 0
}


// Helper function to format dates as 'YYYY-MM-DD'
export function yyyymmdd(date) {
  const d = new Date(date),
        month = '' + (d.getMonth() + 1),
        day = '' + d.getDate(),
        year = d.getFullYear();

  return [year, month.padStart(2, '0'), day.padStart(2, '0')].join('-');
}