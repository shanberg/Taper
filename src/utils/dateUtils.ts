import { TaperDate } from '../TaperDate';

export const isDateThisYear = (date: ScheduleDate): boolean => {
  return date.getFullYear() === new Date().getFullYear();
};

/** Higher-order function to create a cached formatter */
function createCachedFormatter() {
  const dateTimeFormatCache = new Map();

  return function (date: ScheduleDate, lang: string = 'en-US'): LocaleDate {
    const yearFormat = isDateThisYear(date) ? undefined : 'numeric';
    const cacheKey = `${lang}-${yearFormat}`;

    // Check if the formatter is already in the cache
    if (!dateTimeFormatCache.has(cacheKey)) {
      // Create a new formatter and store it in the cache
      const formatter = new Intl.DateTimeFormat(lang, {
        month: 'short',
        day: 'numeric',
        year: yearFormat
      });
      dateTimeFormatCache.set(cacheKey, formatter);
    }

    // Use the cached formatter
    const formatter = dateTimeFormatCache.get(cacheKey);
    return formatter.format(date) as LocaleDate;
  };
}

// Create a cached version of formatDate
export const cachedFormatDate = createCachedFormatter();


export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}
