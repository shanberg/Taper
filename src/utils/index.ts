export * from './dateUtils';
export * from './stepUtils';
export * from './scheduleUtils';
export * from './serializationUtils';
export * from './textUtils';
export * from './languageUtils';