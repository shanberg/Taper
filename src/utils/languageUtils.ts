import { LANGUAGES } from '../consts';

export function getLanguageFromKey(languageKey: string): Language {
  return LANGUAGES.find((l) => l.lang === languageKey) || LANGUAGES[0];
}