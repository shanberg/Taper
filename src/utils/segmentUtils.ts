export const isSegmentInvalid = (segment: Segment): boolean => {
  return segment.dose <= 0 || segment.daysForDose <= 0;
};

/** Returns true if dose and daysForDose are both 0 */
export const isSegmentPlaceholder = (segment: Segment): boolean => {
  if (!segment) return false;
  return segment.dose === 0 && segment.daysForDose === 0;
};

export const sumSegmentsDose = (segments: Segment[]): number => {
  return segments.reduce((sum, segment) => sum + segment.dose * segment.daysForDose, 0);
};

export const sumSegmentsDays = (segments: Segment[]): number => {
  return (
    segments.reduce((sum, segment) => sum + segment.daysForDose, 0)
  );
};

export function isSegmentDirectlyAfterPlaceholder(segments: Segment[], index: number) {
  const segment = segments[index]
  if (!segment) return false;

  if (index === -1) {
    return false;
  }

  const prevSegment = segments[index - 1];
  if (!prevSegment) return false;
  return isSegmentPlaceholder(prevSegment);
}

export function segmentIsOrDirectlyAfterPlaceholder(segments: Segment[], index: number) {
  const segment = segments[index]

  if (!segment) return false;
  if (isSegmentPlaceholder(segment)) {
    return true;
  }
  if (isSegmentDirectlyAfterPlaceholder(segments, index)) {
    return true;
  }
  return false;
}