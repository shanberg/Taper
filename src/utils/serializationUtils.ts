import { TaperDate } from '../TaperDate';

export const serializeSchedule = (schedule: Schedule): SerializedSchedule => {
  return JSON.stringify(schedule, (key, value) => {
    if (key === 'startDate' || key === 'endDate' || key === 'startDateInputValue') {
      return { __brand: 'Date', value: new TaperDate(value).toYYYYMMDD() };
    }
    return value;
  }) as SerializedSchedule;
};

export const deserializeSchedule = (serializedSchedule: SerializedSchedule): Schedule => {
  return JSON.parse(serializedSchedule, (key, value) => {
    if (value && value.__brand === 'Date') {
      return new TaperDate(value.value).toScheduleDate();
    }
    return value;
  }) as Schedule;
};