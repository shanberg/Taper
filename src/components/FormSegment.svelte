<script lang="ts">
	import { createEventDispatcher } from 'svelte';
	import { isSegmentInvalid, isSegmentPlaceholder } from '../utils';
	export let segments: Segment[];
	export let segment: Segment;
	export let index: number;

	const dispatch = createEventDispatcher();
	$: isOnlyRealSegment = segments.length === 2;
	$: isPlaceholder = isSegmentPlaceholder(segment);
	$: isInvalid = !isPlaceholder && isSegmentInvalid(segment);

	function deleteSegmentAtIndex() {
		dispatch('removeSegment', index);
	}
</script>

<tr class="segment" class:isInvalid class:isPlaceholder>
	<td class="dose">
		<label for="dose-{index}">
			<input
				min={1}
				id="dose-{index}"
				step={segment.dose > 5 ? 1 : 0.25}
				type="number"
				inputmode="decimal"
				pattern="[1-9]\d*"
				aria-label="dose-{index}"
				bind:value={segment.dose}
				on:change={() => dispatch('change', segment)}
			/>
		</label>
	</td>
	<td class="days">
		<label for="days-{index}">
			<input
				min={1}
				id="days-{index}"
				type="number"
				inputmode="decimal"
				pattern="[1-9]\d*"
				aria-label="days-{index}"
				bind:value={segment.daysForDose}
				on:change={() => dispatch('change', segment)}
			/>
		</label>
	</td>
	<td class="delete">
		<button
			{...isOnlyRealSegment ? { disabled: true } : {}}
			title="Remove this step"
			class="remove-btn"
			on:click={deleteSegmentAtIndex}>Ã—</button
		>
	</td>
</tr>

<style>

	label {
		display: contents;
	}

	input {
		width: 100%;
		height: 100%;
		border: none;
		outline: none;
	}

	.isPlaceholder {
		color: var(--color-fg-muted);
		user-select: none;

		& input {
			background: transparent;
		}
	}

	.isInvalid td input {
		background: var(--color-status-error-bg);
	}

	.isInvalid:focus-within td input {
		background: var(--color-status-error-bg-muted);
	}

	tr:last-child .remove-btn {
		display: none;
	}

	.remove-btn {
		width: 1.5rem;
		height: 100%;
		border: 0;
		margin-left: 1px;
		border-radius: var(--control-radius);
		transition: all var(--control-transition-duration) ease-in-out;
		color: inherit;
		background: transparent;
		cursor: pointer;
		opacity: 0.25;
		padding: 0;
		text-align: center;
	}

	.remove-btn:disabled {
		cursor: not-allowed;
	}

	/* hide remove button on last segment */

	tr:focus-within .remove-btn,
	tr:hover .remove-btn,
	.remove-btn:focus {
		outline: none;
		opacity: 1;
	}

	tr.isPlaceholder .remove-btn,
	tr.isInvalid .remove-btn,
	.remove-btn:hover:not(:disabled),
	.remove-btn:focus:not(:disabled) {
		opacity: 1;
		color: var(--color-fg-error);
		background: var(--color-status-error-bg-muted);
	}
</style>
