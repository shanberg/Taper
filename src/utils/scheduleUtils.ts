import { TaperDate } from '../TaperDate';
import { TEMPLATES, DEFAULT_LANGUAGE_KEY, DEFAULT_TEMPLATE_KEY } from '../consts';
import { isSegmentInvalid, sumSegmentsDays, sumSegmentsDose, isSegmentPlaceholder } from './segmentUtils';
import { formatSegmentText } from './textUtils';
import { getLanguageFromKey } from './languageUtils';

const PLACEHOLDER_SEGMENT: Segment = { dose: 0, daysForDose: 0 };

export function createInitialSchedule(): Schedule {
  return {
    segments: [...TEMPLATES.Default, PLACEHOLDER_SEGMENT],
    startDate: new TaperDate().toScheduleDate(),
    templateKey: DEFAULT_TEMPLATE_KEY,
    languageKey: DEFAULT_LANGUAGE_KEY,
    displayMode: "segments"
  };
}

export function isValidSchedule(schedule: Schedule): boolean {
  if (!schedule) {
    throw new Error("No schedule provided");
  }
  return !schedule.segments.slice(0, schedule.segments.length - 1).some(s => isSegmentInvalid(s));
}

export function calculateScheduleSummary(schedule: Schedule): string {
  return `${sumSegmentsDose(schedule.segments)}mg over ${sumSegmentsDays(schedule.segments)} days`
};

export function calculateSegmentStartAndEndDates(schedule: Schedule, index: number): SegmentWithStartEndDate {
  const segment = schedule.segments[index];
  const taperStartDate = new TaperDate(schedule.startDate);
  const totalDaysForStartDate =
    schedule.segments
      .slice(0, index)
      .reduce((acc: number, curr: { daysForDose: number }) => acc + curr.daysForDose - 1, 0) +
    index;
  taperStartDate.incrementByDays(totalDaysForStartDate);
  const taperEndDate = new TaperDate(taperStartDate.toScheduleDate());
  taperEndDate.incrementByDays(segment.daysForDose - 1);

  return {
    segment,
    segmentStartDate: taperStartDate.toScheduleDate(),
    segmentEndDate: taperEndDate.toScheduleDate()
  }
}

export function getFormattedListForCopyPaste(schedule: Schedule): string {
  const { segments, languageKey } = schedule;
  const selectedLanguage = getLanguageFromKey(languageKey);

  // Filter out placeholder segments
  const validSegments = segments.filter(segment => !isSegmentPlaceholder(segment));

  // Format each segment
  const formattedSegments = validSegments.map((segment, index) => {
    const { segmentStartDate, segmentEndDate } = calculateSegmentStartAndEndDates(schedule, index);

    return formatSegmentText({
      segment,
      segmentStartDate,
      segmentEndDate,
      index,
      selectedLanguage
    });
  });

  // Combine formatted segments into a single string
  return formattedSegments.join('\n');
}
