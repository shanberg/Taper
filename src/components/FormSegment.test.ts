import { describe, test, expect, vi, beforeEach } from 'vitest';
import { render, fireEvent } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import FormSegment from './FormSegment.svelte'; // Adjust the import path as necessary

// Mock the utility functions
vi.mock('../utils', () => ({
	isSegmentInvalid: vi.fn(),
	isSegmentPlaceholder: vi.fn()
}));

import { isSegmentInvalid, isSegmentPlaceholder } from '../utils';

describe('FormSegment', () => {
	let segments: Segment[];
	let segment: Segment;
	let index: number;

	beforeEach(() => {
		segments = [
			{ dose: 1, daysForDose: 2 },
			{ dose: 3, daysForDose: 4 }
		];
		segment = { dose: 1, daysForDose: 2 };
		index = 0;
		isSegmentInvalid.mockReturnValue(false);
		isSegmentPlaceholder.mockReturnValue(false);
	});

	test('renders correctly', () => {
		const { container } = render(FormSegment, { segments, segment, index });
		expect(container).toMatchSnapshot();
	});

	test('handles dose change', async () => {
		const { getByDisplayValue, component } = render(FormSegment, { segments, segment, index });
		const input = getByDisplayValue('1');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '2' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...segment, dose: 2 }
			})
		);
	});

	test('handles daysForDose change', async () => {
		const { getByDisplayValue, component } = render(FormSegment, { segments, segment, index });
		const input = getByDisplayValue('2');

		const changeHandler = vi.fn();
		component.$on('change', changeHandler);

		await fireEvent.input(input, { target: { value: '3' } });
		await fireEvent.change(input);

		expect(changeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: { ...segment, daysForDose: 3 }
			})
		);
	});

	test('handles remove segment', async () => {
		const { getByTitle, component } = render(FormSegment, { segments, segment, index });
		const button = getByTitle('Remove this step');

		const removeHandler = vi.fn();
		component.$on('removeSegment', removeHandler);

		await fireEvent.click(button);

		expect(removeHandler).toHaveBeenCalledWith(
			expect.objectContaining({
				detail: index
			})
		);
	});

	test('disables remove button when segment is the only real segment', () => {
		segments = [
			{ dose: 1, daysForDose: 1 },
			{ dose: 1, daysForDose: 1 }
		];
		const { getByTitle } = render(FormSegment, { segments, segment, index });
		const button = getByTitle('Remove this step');

		expect(button).toBeDisabled();
	});

	test('applies placeholder class when segment is a placeholder', () => {
		isSegmentPlaceholder.mockReturnValue(true);
		const { container } = render(FormSegment, { segments, segment, index });

		expect(container.querySelector('.segment')).toHaveClass('placeholder');
	});

	test('applies isInvalid class when segment is invalid', () => {
		isSegmentInvalid.mockReturnValue(true);
		const { container } = render(FormSegment, { segments, segment, index });

		expect(container.querySelector('.segment')).toHaveClass('isInvalid');
	});
});
