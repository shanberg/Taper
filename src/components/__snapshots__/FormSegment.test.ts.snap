// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormSegment > renders correctly 1`] = `
<div>
  <div
    class="segment svelte-8qx0cq"
  >
    <div
      class="dose svelte-8qx0cq"
    >
      <label
        class="svelte-8qx0cq"
        for="dose-0"
      >
        <input
          aria-disabled="false"
          aria-label="dose-0"
          class="font_body fs_md bg_bgControl rounded_control border_none w_100% p_0_0.5rem h_controlHeight max-h_controlHeight min-h_controlHeight leading_controlHeight text_inherit ring_none pos_relative focus:z_1 focus:shadow_controlFocusVisible [&.isWarning]:shadow_controlWarning min_1 id_dose-0 step_0.25 inputmode_decimal pattern_[1-9]\\d* aria-label_dose-0"
          id="dose-0"
          inputmode="decimal"
          min="1"
          pattern="[1-9]\\d*"
          placeholder=""
          step="0.25"
          type="number"
        />
      </label>
    </div>
     
    <div
      class="days svelte-8qx0cq"
    >
      <label
        class="svelte-8qx0cq"
        for="days-0"
      >
        <input
          aria-disabled="false"
          aria-label="days-0"
          class="font_body fs_md bg_bgControl rounded_control border_none w_100% p_0_0.5rem h_controlHeight max-h_controlHeight min-h_controlHeight leading_controlHeight text_inherit ring_none pos_relative focus:z_1 focus:shadow_controlFocusVisible [&.isWarning]:shadow_controlWarning min_1 id_days-0 inputmode_decimal pattern_[1-9]\\d* aria-label_days-0"
          id="days-0"
          inputmode="decimal"
          min="1"
          pattern="[1-9]\\d*"
          placeholder=""
          type="number"
        />
      </label>
    </div>
     
    <button
      class="remove-btn svelte-8qx0cq"
      disabled=""
      title="Remove this step"
    >
      Ã—
    </button>
  </div>
</div>
`;
